src/interfaces/file_sink.h

```diff
@@ -0,0 +1,120 @@
+/* -*- c++ -*- */
+/*
+ * Gqrx SDR: Software defined radio receiver powered by GNU Radio and Qt
+ *           https://gqrx.dk/
+ *
+ * Copyright 2021 vladisslav2011@gmail.com.
+ *
+ * Gqrx is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3, or (at your option)
+ * any later version.
+ *
+ * Gqrx is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Gqrx; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#ifndef GQRX_FILE_SINK_C_H
+#define GQRX_FILE_SINK_C_H
+
+#include <gnuradio/blocks/api.h>
+#include <gnuradio/sync_block.h>
+#include <boost/lockfree/queue.hpp>
+
+/*!
+    * \brief Write stream to file without blocking.
+    * \ingroup file_operators_blk
+    */
+class BLOCKS_API file_sink : virtual public gr::sync_block
```

Comment:
I think it would be best to contribute this improvement upstream to GNU Radio rather than having Gqrx maintain its own copy of the File Sink block.

Reply:
Also, I'm not convinced that adding another layer of buffering on top of the standard library & operating system's buffering is a good idea, especially since it requires adding threads and incurring the risk of concurrency bugs.

Reply:
I'll try to contribute this improvement upstream to GNU Radio (after some rework, adding and option switch buffering on/off, an option to allocate static buffer and maybe non-blocking file closing), but I don't think that it will be accepted fast enough... or accepted at all. But I want to use gqrx now and have perfect IQ recordings without skipping.
The standard library does not implement even double buffering and always end up with system write call, that may sometimes block for multiple seconds (on my modern SMR HDDs). Adding some application-controlled buffering always solves this problem at cost of (temporarily) increased memory usage. Writing data in large chunks improves performance and lowers CPU usage.
I have plans to implement buffered file_sourse too as playing back 10...20Msps IQ files, recorded from Hackrf, is problematic with current GNU Radio file_source.

Reply:
> system write call, that may sometimes block for multiple seconds

This sounds like a problem with your operating system. I'd expect write-back caching to solve this problem at the OS layer. Perhaps you need to adjust the cache size?

Reply:
All machines, all operating systems (different versions of Linux) suffer from it more or less. And windows, of course, suffer from it too in a much worse way. I do not own a mac, so maybe mac lives in some different world where there is more RAM, than all running processes (chrome e.g.) could consume and where is always enough free RAM space for disk buffers/cache.

Some fine tuning may help, but, as write system call is blocking by design and can not be made non blocking (like send system call), there is absolutely no guarantee, that it will finish in predictable time. The only way to get predictable low blocking time is application level buffering with large enough buffers. Buffers should be preallocated, as sbrk system call may block or even fail and maybe marked with madvice as being used in near future. 

After turning off swaps (took about 5 minutes) things became a little bit better. Much less skipping in normal conditions.

My test results (hackrf, 20Msps) are the following:
- Current git master: gqrx starts dropping buffers after maximum 10 seconds of recording. The resulting file is unplayable. Constant crackling after 10 seconds.
- iq_tool_fixes: gqrx finishes recording gracefully with error message after 1 minute. The resulting file does not have lost samples. Recording with 16 bits per component (32 bits per sample) proceeds till the end of disk. The resulting file is playable.

Is it better to have a short, but correct file and error message showing, that the target device is too slow or a large file full of garbage?

Reply:
> Is it better to have a short, but correct file and error message showing, that the target device is too slow or a large file full of garbage?

I can believe that some users might prefer the recording to continue, but with dropped samples.

In any case, I still believe that Gqrx is not the right place for this. If you believe this improvement to the File Sink is valuable, please contribute it to GNU Radio. Then Gqrx and all other applications can benefit from it. I do not want to maintain copies of GNU Radio's blocks inside Gqrx.

Reply:
An alternate suggestion: try calling `set_min_output_buffer` on the block preceding the File Sink.

Reply:
src->set_min_output_buffer was the first thing, I have tried. If it is set to ~2 seconds, it still gives choppy sound, interrupted every second. If it is set to more, than 2 seconds ... gqrx crashes with std::bad_alloc.
I've implement simple circular buffer based on std::vector (to prevent memory leaks) and got perfect results. No audio interruptions when the buffer is set to 6-7 seconds and acceptable memory usage. Seek performance is sometimes bad, as well as startup times.
+ I've added current file position readout to keep the slider in sync with actual file position. This prevents the slider runaway when the DSP is stopped, but file playback is not. This closes #1017. I do not see a way to make it work correctly using stock GNU Radio file_source...
As this PR is converted to draft (maybe just close it?), we can continue conversation here: https://github.com/gqrx-sdr/gqrx/pull/1016